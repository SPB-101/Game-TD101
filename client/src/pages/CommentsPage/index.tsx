import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { useTranslation } from "react-i18next";

import { Pagination } from "@component/Pagination";
import { Wrapper } from "@component/Wrapper";
import { CommentsList } from "./CommentsList";
import { CommentForm } from "./CommentForm";

import { COMMENTS_PAGE_LIMIT, COMMENTS_RECORD_LIMIT } from "@constants/index";
import { getCurrentTopicTitle } from "@selectors/widgets/forumPage";
import { getTotal } from "@selectors/widgets/messagesPage";
import { newCurrentPage } from "@thunks/widgets/messages";

import "./style.scss";

import type { Props } from "./types";
import type { State } from "@reducers/index";

export const CommentsBlock = ({ total, title, newCurrentPageThunk }: Props) => {
  const { t } = useTranslation();

  return (
    <>
      <Link to="/forum" className="button button_back forum__button">
        {t("backToForum")}
      </Link>
      <Wrapper className="comments" size={"xl"}>
        <h1 className="comments__title">{title}</h1>
        <div className="comments__list-container">
          <CommentsList className="comments__list" />
        </div>

        <Pagination
          totalRecords={total}
          pageLimit={COMMENTS_PAGE_LIMIT}
          recordLimit={COMMENTS_RECORD_LIMIT}
          onCurrentPage={newCurrentPageThunk}
          className="comments__pagination"
        />

        <CommentForm />
      </Wrapper>
    </>
  );
};

const mapStateToProps = (state: State) => ({
  total: getTotal(state),
  title: getCurrentTopicTitle(state),
});

const mapDispatchToProps = {
  newCurrentPageThunk: newCurrentPage,
};

export const Comments = connect(
  mapStateToProps,
  mapDispatchToProps
)(CommentsBlock);
