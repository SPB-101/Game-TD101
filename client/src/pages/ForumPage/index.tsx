import React, { useCallback, useState } from "react";
import { useTranslation } from "react-i18next";
import classNames from "classnames";
import { Field, Form } from "react-final-form";
import { Link } from "react-router-dom";
import { connect } from "react-redux";

import { Pagination } from "@component/Pagination";
import { Button } from "@component/Button";
import { Wrapper } from "@component/Wrapper";
import { Modal } from "@component/Modal";
import { TextField } from "@component/TextField";
import { Loader } from "@component/Loader";
import { ForumList } from "./ForumList";
import {
  FORUM_RECORD_LIMIT,
  TOPIC_MESSAGES_RECORD_LIMIT,
} from "@constants/index";

import { range, required } from "@utils/validation/rules";
import { validate } from "@utils/validation/validate";
import { State } from "@reducers/index";
import { fetchNewTopicForum, newCurrentPage } from "@thunks/widgets/forum";
import {
  getIsNewTopicLoading,
  getNewTopicError,
  getNewTopicId,
  getTotal,
} from "@selectors/widgets/forumPage";

import "./style.scss";

import type { Props } from "./types";

const rulesFieldsProfile = {
  title: [required, (v: string) => range(v, 4)],
};

export const ForumBlock = ({
  total,
  newCurrentPageThunk,
  fetchNewTopicThunk,
  isNewTopicLoading,
  newTopicErrorMessage,
}: Props) => {
  const [isOpenModal, setIsOpenModal] = useState(false);
  const { t } = useTranslation();

  const createTheme = useCallback((values: Record<string, string>) => {
    fetchNewTopicThunk({
      title: values.title,
    });
  }, []);

  const openModal = useCallback(() => {
    setIsOpenModal(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsOpenModal(false);
  }, []);

  return (
    <>
      <Link to="/menu" className="button button_back forum__button">
        {t("backToMenu")}
      </Link>
      <Wrapper className="forum" size={"xl"}>
        <h1 className="forum__title">{t("forum")}</h1>
        <ul className="forum__list forum__list_border forum-list__header">
          <li className="forum-list__item forum-list__header forum-list__theme forum-list__item_center">
            {t("theme")}
          </li>
          <li className="forum-list__item forum-list__header forum-list__updated">
            {t("lastUpdate")}
          </li>
          <li className="forum-list__item forum-list__header forum-list__comments">
            {t("comments")}
          </li>
        </ul>
        <div className="forum__list-body">
          <ForumList className="forum__list forum__list_column" />
        </div>
        <Pagination
          totalRecords={total}
          pageLimit={FORUM_RECORD_LIMIT}
          recordLimit={TOPIC_MESSAGES_RECORD_LIMIT}
          onCurrentPage={newCurrentPageThunk}
          className="forum__pagination"
        />
        <Button onClick={openModal} className="forum__button">
          {t("createNewTheme")}
        </Button>
      </Wrapper>
      <Modal isOpen={isOpenModal} handleClose={closeModal}>
        <h1 className="forum__title">{t("newTheme")}</h1>

        <Form
          onSubmit={createTheme}
          validate={validate(rulesFieldsProfile)}
          render={({ handleSubmit, form }) => (
            <form
              id="new-topic-form"
              className={classNames("forum__new-theme-form", {
                ["forum__new-theme-form_error"]: newTopicErrorMessage.length,
              })}
              onSubmit={async (event) => {
                await handleSubmit(event);
                form.restart();
              }}
            >
              {isNewTopicLoading && <Loader />}

              {newTopicErrorMessage && (
                <div className="login-page__error-text">
                  <span>{newTopicErrorMessage}</span>
                </div>
              )}

              <Field name="title">
                {({ input, meta }) => (
                  <TextField
                    {...input}
                    error={meta.error && meta.touched ? meta.error : ""}
                    name="title"
                    label={t("themeTitle")}
                    placeholder={t("themeTitle")}
                  />
                )}
              </Field>
              <Button
                type="submit"
                form="new-topic-form"
                disabled={isNewTopicLoading}
                className="forum__button_create"
              >
                {t("createTheme")}
              </Button>
            </form>
          )}
        />
      </Modal>
    </>
  );
};

const mapStateToProps = (state: State) => ({
  total: getTotal(state),
  newTopicId: getNewTopicId(state),
  isNewTopicLoading: getIsNewTopicLoading(state),
  newTopicErrorMessage: getNewTopicError(state),
});

const mapDispatchToProps = {
  newCurrentPageThunk: newCurrentPage,
  fetchNewTopicThunk: fetchNewTopicForum,
};

export const ForumPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(ForumBlock);
