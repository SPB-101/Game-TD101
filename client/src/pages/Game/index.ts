export { GamePage } from "./Game";

import { GameApplication } from "../../game/GameApplication";
export const world: { game: GameApplication | null } = { game: null };
