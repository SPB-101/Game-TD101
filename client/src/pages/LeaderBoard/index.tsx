import React from "react";
import { connect } from "react-redux";

import { Link } from "react-router-dom";
import { useTranslation } from "react-i18next";

import { Wrapper } from "@component/Wrapper";
import { LeaderBoardList } from "./LeaderBoardList";
import { Pagination } from "@component/Pagination";

import {
  LEADERBOARD_PAGE_LIMIT,
  LEADERBOARD_RECORD_LIMIT,
} from "@constants/index";

import { newCurrentPage } from "@thunks/widgets/leaderboard";
import { getIdsLeaderboardCount } from "@selectors/widgets/leaderboardPage";

import type { State } from "@reducers/index";
import type { Props } from "./types";

import "./style.scss";

export const LeaderBoardBlock = ({
  idsLeaderboardCount,
  newCurrentPageThunk,
}: Props) => {
  const { t } = useTranslation();

  return (
    <>
      <Link to="/menu" className="button button_back leader-board__button">
        {t("backToMenu")}
      </Link>
      <Wrapper className={"leader-board"} size={"xl"}>
        <h1 className="leader-board__title">{t("leaderboard")}</h1>
        <LeaderBoardList className="leader-board__list" />
        <Pagination
          className="leader-board__pagination"
          onCurrentPage={newCurrentPageThunk}
          totalRecords={idsLeaderboardCount}
          recordLimit={LEADERBOARD_RECORD_LIMIT}
          pageLimit={LEADERBOARD_PAGE_LIMIT}
        />
      </Wrapper>
    </>
  );
};

const mapStateToProps = (state: State) => ({
  idsLeaderboardCount: getIdsLeaderboardCount(state),
});

const mapDispatchToProps = {
  newCurrentPageThunk: newCurrentPage,
};

export const LeaderBoardPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(LeaderBoardBlock);
