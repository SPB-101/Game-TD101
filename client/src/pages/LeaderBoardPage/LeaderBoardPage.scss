.leader-board {
  display: flex;
  height: 100vh;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  color: var(--nord0);

  &__title {
    font-size: var(--text-size-1);
    text-shadow: var(--shadow-glitch);
    text-transform: uppercase;
    margin-bottom: 35px;
  }

  &__list {
    width: 50%;
    margin-bottom: 42px;
  }
}
