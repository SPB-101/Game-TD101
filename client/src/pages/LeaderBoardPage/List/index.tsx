import React, { useEffect } from "react";
import { connect } from "react-redux";

import { useTranslation } from "react-i18next";

import { Loader } from "@component/Loader";
import { Pagination } from "@component/Pagination";
import { Item } from "../Item";

import { getIsLoading } from "@selectors/widgets/leaderboardPage";
import {
  getCurrentPageLeaderboard,
  getIdsLeaderboardCount,
} from "@selectors/widgets/leaderboardPage";
import { fetchLeaderboard } from "@thunks/collections/leaderboard";
import { newCurrentPage } from "@thunks/widgets/leaderboard";

import {
  LEADERBOARD_TAG,
  LEADERBOARD_PAGE_LIMIT,
  LEADERBOARD_RECORD_LIMIT,
} from "@constants/index";

import type { State } from "@reducers/index";
import type { Props } from "./types";

import "./style.css";

export const ListBlock = ({
  className,
  isLoading,
  idsLeaderboard,
  idsLeaderboardCount,
  fetchLeaderboardThunk,
  newCurrentPageThunk,
}: Props): JSX.Element => {
  const { t } = useTranslation();
  useEffect(() => {
    fetchLeaderboardThunk({
      ratingFieldName: LEADERBOARD_TAG,
      cursor: 0,
      limit: LEADERBOARD_RECORD_LIMIT * 20,
    });
  }, []);

  return (
    <>
      <ul className={className}>
        {isLoading ? (
          <Loader />
        ) : idsLeaderboard.length ? (
          idsLeaderboard.map((id, i) => {
            return <Item key={id} id={id} index={i} />;
          })
        ) : (
          t("emptyLeaderboard")
        )}
      </ul>
      <Pagination
        onCurrentPage={newCurrentPageThunk}
        totalRecords={idsLeaderboardCount}
        recordLimit={LEADERBOARD_RECORD_LIMIT}
        pageLimit={LEADERBOARD_PAGE_LIMIT}
      />
    </>
  );
};

const mapStateToProps = (state: State) => ({
  isLoading: getIsLoading(state),
  idsLeaderboard: getCurrentPageLeaderboard(state),
  idsLeaderboardCount: getIdsLeaderboardCount(state),
});

const mapDispatchToProps = {
  fetchLeaderboardThunk: fetchLeaderboard,
  newCurrentPageThunk: newCurrentPage,
};

export const List = connect(mapStateToProps, mapDispatchToProps)(ListBlock);
