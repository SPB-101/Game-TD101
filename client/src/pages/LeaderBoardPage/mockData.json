[
  {
    "name": "Kristina Trifunov",
    "score": 12123232,
    "photo": "https://uifaces.co/our-content/donated/7JZz4G7I.jpg"
  },
  {
    "name": "Vivian Painter",
    "score": 12323,
    "photo": "https://images.generated.photos/QbcqC7vbdYI9lI-q6inuEaCPd_yC4v7OkPMqCdWk-tM/rs:fit:512:512/Z3M6Ly9nZW5lcmF0/ZWQtcGhvdG9zLzA5/OTk2NjMuanBn.jpg"
  },
  {
    "name": "Aniah Lassiter",
    "score": 9898,
    "photo": "https://images.generated.photos/rF3Ir6Y3AZiiqNshk_D1rTEAUsgYz4wv_CLWd0EJZmM/rs:fit:512:512/Z3M6Ly9nZW5lcmF0/ZWQtcGhvdG9zL3Yy/XzA1NzM2MTUuanBn.jpg"
  },
  {
    "name": "Milagros Betts",
    "score": 234,
    "photo": "https://i.imgur.com/h2E4WGw.jpg"
  },
  {
    "name": "Krish Kendall",
    "score": 3490,
    "photo": "https://images.generated.photos/KSkFZay8lawllgpfawH1UjaofmrXm1sPLZiauTWaYag/rs:fit:512:512/Z3M6Ly9nZW5lcmF0/ZWQtcGhvdG9zLzAw/MzYyNTIuanBn.jpg"
  },
  {
    "name": "Dave Smiths",
    "score": 1233
  }
]
