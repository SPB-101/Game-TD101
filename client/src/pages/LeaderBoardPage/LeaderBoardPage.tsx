import React, { useEffect } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { useTranslation } from "react-i18next";

import { LeaderBoardItem } from "../../component/LeaderBoardItem";
import { Pagination } from "../../component/Pagination";
import { Wrapper } from "../../component/Wrapper";
import { Loader } from "../../component/Loader";

import { getIsLoading } from "../../store/selectors/widgets/leaderboardPage";
import { getLeaderboard } from "../../store/selectors/collections/leaderboard";
import { fetchLeaderboard } from "../../store/thunks/collections/leaderboard";

import { LEADERBOARD_TAG } from "../../constants";
import type { State } from "../../store/reducers";
import type { Props } from "./types";

import "./LeaderBoardPage.scss";

export const LeaderBoardBlock = ({
  isLoading,
  leaderboard,
  fetchLeaderboardThunk,
}: Props): JSX.Element => {
  const { t } = useTranslation();
  useEffect(() => {
    fetchLeaderboardThunk({
      ratingFieldName: LEADERBOARD_TAG,
      cursor: 0,
      limit: 5,
    });
  }, []);

  return (
    <>
      <Link to="/menu" className="button button_back leader-board__button">
        {t("backToMenu")}
      </Link>
      <Wrapper className={"leader-board"} size={"xl"}>
        {isLoading ? <Loader /> : null}
        <h1 className="leader-board__title">{t("leaderboard")}</h1>
        <ul className="leader-board__list">
          {leaderboard.length === 0 && !isLoading
            ? t("emptyLeaderboard")
            : leaderboard.map((item, index) => {
                return (
                  <LeaderBoardItem
                    key={index}
                    rankingPosition={index + 1}
                    playerName={item.displayName}
                    score={item[LEADERBOARD_TAG]}
                    playerImage={item.avatar ? item.avatar : ""}
                  />
                );
              })}
        </ul>
        <Pagination totalRecords={leaderboard.length} pageLimit={5} />
      </Wrapper>
    </>
  );
};

const mapStateToProps = (state: State) => ({
  isLoading: getIsLoading(state),
  leaderboard: getLeaderboard(state),
});

export const LeaderBoardPage = connect(mapStateToProps, {
  fetchLeaderboardThunk: fetchLeaderboard,
})(LeaderBoardBlock);
