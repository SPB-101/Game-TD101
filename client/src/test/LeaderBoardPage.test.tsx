import React from "react";
import renderer from "react-test-renderer";
import { LeaderBoardPage } from "../pages/LeaderBoardPage";
import { StaticRouter } from "react-router";

describe("<LeaderBoardPage />", () => {
  it("LeaderBoardPage renders correctly", () => {
    const tree = renderer
      .create(
        <StaticRouter location="/leaderboard">
          <LeaderBoardPage />
        </StaticRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
