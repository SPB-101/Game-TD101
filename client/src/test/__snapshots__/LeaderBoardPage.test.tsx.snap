// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LeaderBoardPage /> LeaderBoardPage renders correctly 1`] = `
Array [
  <a
    className="button button_back leader-board__button"
    href="/menu"
    onClick={[Function]}
  >
    backToMenu
  </a>,
  <div
    className="wrapper wrapper_size--xl leader-board"
  >
    <h1
      className="leader-board__title"
    >
      leaderboard
    </h1>
    <ul
      className="leader-board__list"
    >
      <li
        className="leader-board__item"
      >
        <div
          className="item__ranking"
        >
          <winner-icon.svg
            height="50px"
            width="50px"
          />
          <span
            className="item_color--grey item_position--absolute item_z-index-low"
          >
            1
          </span>
        </div>
        <div
          className="item__player"
        >
          <img
            alt="Kristina Trifunov's photo"
            className="avatar item__player-image"
            height={50}
            src="https://uifaces.co/our-content/donated/7JZz4G7I.jpg"
            width={50}
          />
          <span
            className="item__player-name"
          >
            Kristina Trifunov
          </span>
        </div>
        <span
          className="item__score"
        >
          12123232
        </span>
      </li>
      <li
        className="leader-board__item"
      >
        <div
          className="item__ranking"
        >
          <leader-icon.svg
            fill="#f2f2f2"
            height="50px"
            width="50px"
          />
          <span
            className="item_color--grey item_position--absolute"
          >
            2
          </span>
        </div>
        <div
          className="item__player"
        >
          <img
            alt="Vivian Painter's photo"
            className="avatar item__player-image"
            height={50}
            src="https://images.generated.photos/QbcqC7vbdYI9lI-q6inuEaCPd_yC4v7OkPMqCdWk-tM/rs:fit:512:512/Z3M6Ly9nZW5lcmF0/ZWQtcGhvdG9zLzA5/OTk2NjMuanBn.jpg"
            width={50}
          />
          <span
            className="item__player-name"
          >
            Vivian Painter
          </span>
        </div>
        <span
          className="item__score"
        >
          12323
        </span>
      </li>
      <li
        className="leader-board__item"
      >
        <div
          className="item__ranking"
        >
          <leader-icon.svg
            fill="#f2994a"
            height="50px"
            width="50px"
          />
          <span
            className="item_color--grey item_position--absolute"
          >
            3
          </span>
        </div>
        <div
          className="item__player"
        >
          <img
            alt="Aniah Lassiter's photo"
            className="avatar item__player-image"
            height={50}
            src="https://images.generated.photos/rF3Ir6Y3AZiiqNshk_D1rTEAUsgYz4wv_CLWd0EJZmM/rs:fit:512:512/Z3M6Ly9nZW5lcmF0/ZWQtcGhvdG9zL3Yy/XzA1NzM2MTUuanBn.jpg"
            width={50}
          />
          <span
            className="item__player-name"
          >
            Aniah Lassiter
          </span>
        </div>
        <span
          className="item__score"
        >
          9898
        </span>
      </li>
      <li
        className="leader-board__item"
      >
        <div
          className="item__ranking"
        >
          <span
            className="item__ranking-number"
          >
            4
          </span>
        </div>
        <div
          className="item__player"
        >
          <img
            alt="Krish Kendall's photo"
            className="avatar item__player-image"
            height={50}
            src="https://images.generated.photos/KSkFZay8lawllgpfawH1UjaofmrXm1sPLZiauTWaYag/rs:fit:512:512/Z3M6Ly9nZW5lcmF0/ZWQtcGhvdG9zLzAw/MzYyNTIuanBn.jpg"
            width={50}
          />
          <span
            className="item__player-name"
          >
            Krish Kendall
          </span>
        </div>
        <span
          className="item__score"
        >
          3490
        </span>
      </li>
      <li
        className="leader-board__item"
      >
        <div
          className="item__ranking"
        >
          <span
            className="item__ranking-number"
          >
            5
          </span>
        </div>
        <div
          className="item__player"
        >
          <profile-icon.svg
            className="avatar item__player-image"
            height="50px"
            width="50px"
          />
          <span
            className="item__player-name"
          >
            Dave Smiths
          </span>
        </div>
        <span
          className="item__score"
        >
          1233
        </span>
      </li>
    </ul>
    <ul
      className="pagination"
    >
      <li
        className="pagination__item"
      >
        <a
          className="arrow arrow--prev"
          onClick={[Function]}
        />
      </li>
      <li
        className="pagination__item"
      >
        <a
          className="pagination__link active"
          onClick={[Function]}
        >
          1
        </a>
      </li>
      <li
        className="pagination__item"
      >
        <a
          className="pagination__link"
          onClick={[Function]}
        >
          2
        </a>
      </li>
      <li
        className="pagination__item"
      >
        <a
          className="pagination__link"
          onClick={[Function]}
        >
          3
        </a>
      </li>
      <li
        className="pagination__item"
      >
        <a
          className="arrow arrow--next"
          onClick={[Function]}
        />
      </li>
    </ul>
  </div>,
]
`;
